byway/src/main/webapp/routepage.js

```diff
@@ -49,27 +63,139 @@ function calcRoute(directionsService, directionsRenderer, start, end) {
+      });
+     stopList.appendChild(btn);
+    })
+  })
+}
+
+/** Get the new list of stops locally */
+function getStopsList(){
+  const stopList = document.getElementById('stop-list');
+
+  if(stopList != null){
+    stopList.innerText = ""; 
+  }
+
+  // re-render list synchronously
+  stops.forEach((stop)=>{
+    var btn = document.createElement('button');
+    btn.innerText = stop.placename
+    btn.setAttribute("class", "btn rec-btn");
+    btn.addEventListener("click", function() {
+      deleteFromStops(stop); 
+    });
+    stopList.appendChild(btn);
+  })
+}
+
+/** Add stop locally and to datastore */
 function addToStops(stop){
-  console.log("add" + stop.placename + "to the ArrayList in the servlet");
+  deleteFromRecs(stop); 
+
+  // add to stops array locally in js
+  stops.push(stop);
+  getStopsList();
+
+  // add to datastore
+  const params = new URLSearchParams();
+  params.append("text", stop.placename);
+  params.append("action", "add");
+  fetch('/api/stop', {method: 'POST', body: params});
 }
 
+/** Delete stop locally and from datastore*/
+function deleteFromStops(stop){
+  addToRecs(stop); 
+
+  // delete from stops array locally in js
+  stops = stops.filter(function(stopObj){
+    return stopObj.placename != stop.placename;
+  })
+  getStopsList();
+
+  // delete from datastore
+  const params = new URLSearchParams();
+  params.append("text", stop.placename);
+  params.append("action", "remove");
+  fetch('/api/stop', {method: 'POST', body: params});
+}
+
+/** Get the new list of recommendations from servlet onload */
 function getRecs() {
+  const recList = document.getElementById('rec-list');
+  if(recList != null){
+    recList.innerText = ""; // clear list
+  }
   fetch('/api/recs')
   .then(response => response.json())
-  .then((recs) => {
-    const recList = document.getElementById('rec-list');
-    recList.innerText = ""; // clear old comments 
-    recs.forEach((stop) => {
-      let btn = document.createElement("button");
-      btn.innerText = stop.placename;
+  .then((recommendations) => {
+    recommendations.forEach((rec)=>{
+      // populate local js array initially
+      recs.push(rec);
+      var btn = document.createElement('button');
+      btn.id = rec.id;
+      btn.innerText = rec.placename;
       btn.setAttribute("class", "btn rec-btn");
       btn.addEventListener("click", function() {
-        addToStops(stop);
+        addToStops(rec); 
       });
-      recList.appendChild(btn);
+     recList.appendChild(btn);
+    })
+  })
+}
+
+/** Get the new list of recommendations locally */
+function getRecsList(){
+  const recsList = document.getElementById('rec-list');
+  recsList.innerText = ""; // clear list
+  // re-render list synchronously
+  recs.forEach((rec)=>{
+    var btn = document.createElement('button');
+    btn.innerText = rec.placename
+    btn.setAttribute("class", "btn rec-btn");
+    btn.addEventListener("click", function() {
+      addToStops(rec);
     });
```

Comment:
Similarly, can decompose since this is a similar functionality to what you're doing in `getRecs()`

Reply:
Done!